### PR TITLE
Make minkelite more robust for multi client use.

### PR DESCRIPTION
**Pull request by Setogit**
Add over_write config parameter and change the test to use it.
Use moe universal db file name in case the client requests
to use a default name.

connect strongloop-internal/scrum-nodeops#843

### CODE REVIEW

**Comment by Setogit** [comment]
connect https://github.com/strongloop-internal/scrum-nodeops/issues/843

**Comment by rmg** [comment]
overwrite is one word :-)

**Comment by sam-github** [comment]
Maybe I'm missing it, but this PR seems to:

1. use a different name for the DB every time minkelite/strong-pm starts
2. delete the old DB (except it won't, because it doesn't know the old DB name, see above)

So, it will leak old DBs.

If you are going to delete the old DB every time on startup, why not always use the same DB name?

Also, none of these changes seem to address the issues described by you [here](https://github.com/strongloop-internal/scrum-nodeops/issues/843#issuecomment-131004447). Why would any one of those conditions be true?

File: index.js [hunk 1]
```diff
@@ -175,9 +202,15 @@ MinkeLite.prototype._init_db = function () {
     false :
     fs.existsSync(this.db_path)
 
+  if(this.db_exists && this.config.over_write) {
+    fs.unlinkSync(this.db_path);
```
Comment (sam-github) [inline index.js]: This is a race condition: if the DB is removed between the check of its existence above, and the attempt to unlink here, then the `unlinkSync()` will throw an error. You should wrap the unlink with try-catch.

Use of the `exists()` is often an anti-pattern for this reason, its almost always racy. Its best to attempt to use a resource and handle its existence/non-existence than to check if it exists at some point in time, and assume that state hasn't changed at some later time.

**Comment by sam-github** [comment]
Can you reproduce this issue?

**Comment by rmg** [comment]
If the intention really is to give each run a unique database, then it would be simpler to just pass an empty string as the file name. That tells sqlite to create an anonymous file backed database that is deleted once it is closed. Then you don't have to worry about making a unique file name.

But like @sam-github said, I'm not sure if that's actually the goal.

File: index.js [hunk 2]
```diff
@@ -175,9 +202,15 @@ MinkeLite.prototype._init_db = function () {
...
+    this.db_exists = false;
+  }
+
   debug('db open: path %j pre-existing? %j', this.db_path, this.db_exists);
 
-  this.db = new sqlite3.Database(this.db_path)
+  this.db = new sqlite3.Database(this.db_path,
+      sqlite3.OPEN_READWRITE | sqlite3.OPEN_CREATE)
```
Comment (rmg) [inline index.js]: These are the default flags, so this shouldn't be necessary.

File: index.js (hunk 1 above)
Comment (Setogit) [inline index.js]: Wrapped with try-catch.

**Comment by Setogit** [comment]
Made it one word "overwrite" :-)  Wrapped unlinkSync with try-catch.

Reason of the default db file name is the following:  When the name is specified by the client with overwrite true or false, it is left to the client.  The client can do anything on the file access right or even the contents.  Only change made in this PR from this point of view is to explicitly set the mode to sqlite3.OPEN_READWRITE | sqlite3.OPEN_CREATE.

In the default case: no name is specified by the client, giving it an unique name every time is to save the db file for post incident analysis. Minkelite is designed to support this scenario: you save the DB file and use it for offline analysis by openin it with pruning_interval_seconds=0 and stale_minutes=0 (minkelite stops auto-pruning).  You can use the same UI to do the offline analysis as long as you'd like.

However, that feature is not advertised.  In other words, the value of offline analysis not recognized.  It may make more sense to leave it sqlite3 to use anonymous name and delete it when closed.

File: index.js (hunk 2 above)
Comment (Setogit) [inline index.js]: Was relying on the default.  I prefer to explicitly set it this time.

File: README.md [hunk 3]
```diff
@@ -123,5 +123,6 @@ Returns an Express app of the MinkeLite instance.  Clients should start it.
 ### ml.startServer()
 Starts the Express app of the MinkeLite instance
 
-### ml.shutdown()
+### ml.shutdown(callback)
 Shutdown the handles created for the MinkeLite instance
+- `callback` {Function} Calleld with error
```
Comment (sam-github) [inline README.md]: "Called"

File: index.js [hunk 4]
```diff
@@ -95,7 +113,7 @@ function MinkeLite(config) {
 
   this.config.dir_path = this.config.dir_path || "./"
   this.config.db_name = this.config.in_memory ? ":memory:" : (
-    this.config.db_name || "minkelite.db" );
+    this.config.db_name || '' );
```
Comment (sam-github) [inline index.js]: `If the filename is an empty string, then a private, temporary on-disk database will be created. This private database will be automatically deleted as soon as the database connection is closed.`  https://www.sqlite.org/c3ref/open.html

If it doesn't get closed, if node exist with SIGTERM for example, what then?

**Comment by sam-github** [comment]
@slnode test please

**Comment by sam-github** [comment]
LGTM, you can do a command line merge when the tests are green.

File: index.js (hunk 4 above)
Comment (Setogit) [inline index.js]: https://www.sqlite.org/tempfiles.html discussed temp file usage in sqlite, but it's not clear what happens if node exits without closing the connection.  Wrote a test code for these 3 cases:

`var params = [
	{testId: 'In-Memory', data: {'in_memory': true, 'db_name': 'minkelite_test.db', 'overwrite': true}},
	{testId: 'On-File/Named', data: {'in_memory': false, 'db_name': 'minkelite_test.db', 'overwrite': true}},
	{testId: 'On-File/Anonymous', data: {'in_memory': false, 'db_name': '', 'overwrite': true}}
];`

**Comment by sam-github** [comment]
Can you rebase? Your first commit adds code that the second and fourth commit remove, and the fifth commit tests:
```
83fb8b3 Add test to cover three sqlite3 access modes
6eca04b Remove redundant code
9867162 Let sqlite3 use an anonymous file name
8e1e79d Make minkelite more robust for multi client use.
^^^^ above four should be one commit, with a title that says what it does, "support a temporary on-disk DB"
e907897 Fix up description of shudown in README
^^^^ the commit message has a typo
```

**Comment by sam-github** [comment]
LGTM, other than needing rebasing and squashing

**Comment by sam-github** [comment]
FYI, @Setogit @rmg, Richard Hipp says:

**Comment by sam-github** [comment]
Needs verification that traces is still working as expected.

Doesn't need long running, but it should be checked with pm killed and restarted. With the new change, after SIGKILL, the DB should re-open, but there will be no tracing data in minkelite from before the pm death (but that has always been true because arc doesn't show dead pids).


@cgole

**Pull request merged**